lib/active_record/connection_adapters/fibered_mysql2_adapter.rb

```diff
@@ -1,82 +1,72 @@
 # frozen_string_literal: true
 
-require 'em-synchrony'
 require 'active_model'
 require 'active_record/errors'
-require 'active_record/connection_adapters/em_mysql2_adapter'
+require 'active_record/connection_adapters/mysql2_adapter'
+require_relative '../../fibered_mysql2/async_task'
 
 module FiberedMysql2
-  module FiberedMysql2Adapter_4_2
+  module FiberedMysql2Adapter_6
     def lease
-      synchronize do
-        unless in_use?
-          @owner = Fiber.current
-        end
-      end
-    end
-  end
-
-  module FiberedMysql2Adapter_5_2
-    def lease
-      if in_use?
-        msg = "Cannot lease connection, ".dup
-        if owner_fiber == Fiber.current
-          msg << "it is already leased by the current fiber."
+      if (ot = owner_task)
+        msg = +"Cannot lease connection; "
+        if ot == (current_task = AsyncTask.current_or_none)
+          msg << "it is already leased by the current Async::Task."
         else
-          msg << "it is already in use by a different fiber: #{owner_fiber}. " \
-                  "Current fiber: #{Fiber.current}."
+          msg << "it is already in use by a different Async::Task: #{ot}. " \
+                  "Current Async::Task: #{current_task}."
         end
         raise ::ActiveRecord::ActiveRecordError, msg
       end
 
-      @owner = Fiber.current
+      @owner = AsyncTask.current_or_none
     end
 
     def expire
-      if in_use?
-        # Because we are actively releasing connections from dead fibers, we only want
-        # to enforce that we're expiring the current fibers connection, iff the owner
+      if (ot = owner_task)
```

Comment:
Looking at the `owner_task` as defined below

```
def owner_task
  @owner.nil? || @owner.is_a?(Async::Task) or
    raise "@owner must be an Async::Task! Found #{@owner.inspect}"
   
  @owner
end
```

`@owner` could be `nil` isn't it? In which case all the subsequent calls on it like `alive?` could fail isn't it?

So we may need to add another check for not nil like `!ot.nil? && ot.alive? ...` ?

Reply:
But short-circuiting `||` means that if it's `nil`, we won't check `alive?`, right?

Reply:
Also as mentioned in Slack, I'm leaning back to using `Fiber` as the owning object. Even though `Async::Task` seems more precise and less subject to corner-cases in Ruby... but as soon as we move to Rails 7, we're going to be using `isolation_level: :fiber`, so we might as well now.

Reply:
^ Done.